Arm NEON thumb support

emesare opened the pull request.

emesare committed:
Remove duplicate conditional checks in v_ lifts

emesare committed:
Fix getting incorrect S and D registers in lifted il

emesare committed:
Lift VPOP and VPUSH

emesare committed:
Lift VLDR and VSTR

emesare committed:
Lift VMOV

emesare committed:
Lift misc V_ thumb instrunctions

emesare committed:
Fix non R register groups referencing PC for 15th register index

Any usage of d15, s15, q15 was completely broken

emesare committed:
Lift VSTMIA, VLDMIA

emesare committed:
Lift more variants of VCVT and correct a missed variant case

emesare committed:
Lift VSUB, VDIV, VMUL, VFMA

VSUB already existed but it was lifted as a regular SUB instruction

emesare committed:
Lift VMRS and mark system registers

System registers needed to be marked to prevent the registers from being passed as arguments

emesare committed:
Fix unconstrained VADD and VABS

emesare committed:
Fix VDIV constraints and lift VNEG

emesare committed:
Fix VPOP not setting registers

emesare committed:
Use split registers for VMOV

emesare committed:
Constraint NEON instructions using operation flags

Also fix some formatting issues

emesare commented:
Should be well formatted now, if anyone has opinions on the formatting of calls they can push their formatting on top of this branch.

galenbwill commented on arch/armv7/thumb2_disasm/il_thumb2.cpp:
```diff
@@ -1475,7 +1512,289 @@ bool GetLowLevelILForNEONInstruction(Architecture* arch, LowLevelILFunction& il,
+				// r3:r2 <- d12
+				il.SetRegisterSplit(4, ReadILOperand(il, instr, 0), ReadILOperand(il, instr, 1),
+				                    ReadILOperand(il, instr, 3));
+			} else
```
`else` should be on its own line

galenbwill commented on arch/armv7/thumb2_disasm/il_thumb2.cpp:
```diff
@@ -1475,7 +1512,289 @@ bool GetLowLevelILForNEONInstruction(Architecture* arch, LowLevelILFunction& il,
+		}
+		break;
+	case armv7::ARMV7_VSTR:
+		{
```
don't indent this `{`, or the closing `}`, but do move the `break;` inside the curly braces so it is indented

galenbwill commented on arch/armv7/thumb2_disasm/il_thumb2.cpp:
```diff
@@ -1475,7 +1512,289 @@ bool GetLowLevelILForNEONInstruction(Architecture* arch, LowLevelILFunction& il,
+		if(IS_FIELD_PRESENT(instr, FIELD_inc))
+			inc = instr->fields[FIELD_inc];
+		int regs = instr->fields[FIELD_regs];
+		for(int i=0; i<regs; ++i) {
```
spaces around binary operators, `{` on its own line

galenbwill commented on arch/armv7/thumb2_disasm/il_thumb2.cpp:
```diff
@@ -1475,7 +1512,289 @@ bool GetLowLevelILForNEONInstruction(Architecture* arch, LowLevelILFunction& il,
+		auto regSize = RegisterSizeFromPrefix(prefix);
+		unsigned int d = instr->fields[FIELD_d];
+		unsigned int inc = 1;
+		if(IS_FIELD_PRESENT(instr, FIELD_inc))
```
space after `if`

galenbwill commented on arch/armv7/thumb2_disasm/il_thumb2.cpp:
```diff
@@ -1475,7 +1512,289 @@ bool GetLowLevelILForNEONInstruction(Architecture* arch, LowLevelILFunction& il,
+		{
+		// TODO: Clean this code up...
+		const char* prefix = "d";
+		if (IS_FIELD_PRESENT(instr, FIELD_single_regs)){
```
newline before `{`

galenbwill committed:
Cleaned up formatting on new and old code

galenbwill committed:
Fixed fall-through VMOV lifting

Conditions were not cascading, caused instructions to generate il for multiple variants of VMOV
